batch/batch/worker/worker.py

```diff
@@ -1502,31 +1505,46 @@ async def run(self):
                 if not user_error(e):
                     log.exception(f'while running {self}')
 
-                self.state = 'error'
-                self.error = traceback.format_exc()
+                if isinstance(e, JobDeletedError):
+                    self.state = 'cancelled'
+                else:
+                    self.state = 'error'
+                    self.error = traceback.format_exc()
             finally:
                 with self.step('post-job finally block'):
-                    if self.disk:
-                        try:
-                            await self.disk.delete()
-                            log.info(f'deleted disk {self.disk.name} for {self.id}')
-                        except asyncio.CancelledError:
-                            raise
-                        except Exception:
-                            log.exception(f'while detaching and deleting disk {self.disk.name} for {self.id}')
-                        finally:
-                            await self.disk.close()
-                    else:
-                        self.worker.data_disk_space_remaining.value += self.external_storage_in_gib
-
-                    await self.cleanup()
+                    try:
+                        if self.disk:
+                            try:
+                                await self.disk.delete()
+                                log.info(f'deleted disk {self.disk.name} for {self.id}')
+                            except asyncio.CancelledError:
+                                raise
+                            except Exception:
+                                log.exception(f'while detaching and deleting disk {self.disk.name} for {self.id}')
+                            finally:
+                                await self.disk.close()
+                        else:
+                            self.worker.data_disk_space_remaining.value += self.external_storage_in_gib
+                    finally:
```

Comment:
It seems like an ExitStack is the right move here.

```python3
async def cleanup(self):
    async with AsyncExitStack() as stack:
        if self.disk:
            stack.push_async_callback(self.disk.delete)
            stack.push_async_callback(self.disk.close)
        async def umount_cloudfuse(config):
            bucket = config['bucket']
            assert bucket
            mount_path = self.cloudfuse_data_path(bucket)
            await CLOUD_WORKER_API.unmount_cloudfuse(mount_path)
            config['mounted'] = False

        for config in self.cloudfuse or []:
            if config['mounted']:
                stack.push_async_callback(umount_cloudfuse, config)
        stack.push_async_callback(check_shell, f'xfs_quota -x -c "limit -p bsoft=0 bhard=0 {self.project_id}" /host')
        stack.push_async_callback(blocking_to_async, self.pool, shutil.rmtree, self.scratch, ignore_errors=True)
    log.info(f'successfully cleaned up {self}')
```

Any error which propagates out of cleanup will go to run which will go to run_job which will log appropriately.

If it fails you get a nicely nested exception. Every function will be run, even if previous ones fail.
```python3
In [6]: def boom(x): 
   ...:     raise ValueError(x) 
   ...:  
   ...: with ExitStack() as stack: 
   ...:     stack.callback(boom, 1) 
   ...:     stack.callback(boom, 2) 
   ...:     stack.callback(boom, 3) 
   ...:     stack.callback(boom, 4)     
   ...:     stack.callback(print, "hello") 
   ...:                                                                                                                                                        
hello
Traceback (most recent call last):
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 509, in __exit__
    if cb(*exc_details):
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 384, in _exit_wrapper
    callback(*args, **kwds)
  File "<ipython-input-6-64a8daeefc62>", line 2, in boom
    raise ValueError(x)
ValueError: 4

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 509, in __exit__
    if cb(*exc_details):
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 384, in _exit_wrapper
    callback(*args, **kwds)
  File "<ipython-input-6-64a8daeefc62>", line 2, in boom
    raise ValueError(x)
ValueError: 3

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 509, in __exit__
    if cb(*exc_details):
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 384, in _exit_wrapper
    callback(*args, **kwds)
  File "<ipython-input-6-64a8daeefc62>", line 2, in boom
    raise ValueError(x)
ValueError: 2

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<ipython-input-6-64a8daeefc62>", line 9, in <module>
    stack.callback(print, "hello")
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 524, in __exit__
    raise exc_details[1]
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 509, in __exit__
    if cb(*exc_details):
  File "/Users/dking/miniconda3/lib/python3.7/contextlib.py", line 384, in _exit_wrapper
    callback(*args, **kwds)
  File "<ipython-input-6-64a8daeefc62>", line 2, in boom
    raise ValueError(x)
ValueError: 1
```

Reply:
I'm a bit uncomfortable with this change because it will change the logging messages. I'll make the correctness change and we can add exit stacks later.

Reply:
I refactored the cleanup code a bit as there was no reason for separating the steps into two places